### PR TITLE
Fixed #30318 --  Added check for importability of arguments of custom error handler views.

### PR DESCRIPTION
**Pull request by alasdairnicol**
…mported

Thanks to Jon on Stack Overflow for reporting the issue.

### CODE REVIEW

File: django/urls/resolvers.py [hunk 1]
```diff
@@ -405,7 +405,15 @@ def _check_custom_error_handlers(self):
...
         for status_code, num_parameters in [(400, 2), (403, 2), (404, 2), (500, 1)]:
-            handler, param_dict = self.resolve_error_handler(status_code)
+            try:
+                handler, param_dict = self.resolve_error_handler(status_code)
+            except (ModuleNotFoundError, ViewDoesNotExist) as e:
+                msg = (
+                    "The custom handler{status_code} view could not be imported."
+
+                ).format(status_code=status_code)
+                messages.append(Error(msg, hint=e.args[0], id='urls.E008'))
```
Comment (alasdairnicol) [inline django/urls/resolvers.py]: I'm not sure that `e.args[0]` is the best way to get the message. `e.msg` works for `ModuleNotFound` but not `ViewDoesNotExist`.

File: django/urls/resolvers.py [hunk 2]
```diff
@@ -405,7 +405,15 @@ def _check_custom_error_handlers(self):
         # All handlers take (request, exception) arguments except handler500
         # which takes (request).
         for status_code, num_parameters in [(400, 2), (403, 2), (404, 2), (500, 1)]:
-            handler, param_dict = self.resolve_error_handler(status_code)
+            try:
+                handler, param_dict = self.resolve_error_handler(status_code)
+            except (ModuleNotFoundError, ViewDoesNotExist) as e:
+                msg = (
+                    "The custom handler{status_code} view could not be imported."
```
Comment (alasdairnicol) [inline django/urls/resolvers.py]: Initially I wanted the message to be "The custom handler{status_code} 'path.to.view' could not be imported", but I couldn't see a way to get `path.to.view` without duplicating `resolve_error_handler`, or extracting all but the final line of `resolve_error_handler` into a new method. I think that including the exception message in the hint might be enough information.

**Comment by alasdairnicol** [comment]
Should we add a note to the 2.2.1 release notes, and if so, should we add a note about the urls.E007 error to the 2.2 release notes as well?

**Comment by alasdairnicol** [comment]
@carltongibson thanks for the review. Nice idea to use `getattr`, that works nicely.

Including the traceback message in the hint gives a bit more information that might be helpful (e.g. whether it's the module or method that is invalid), but I can remove that if you want.

I've added a note to the release notes. I'm not sure if it's fair for me to call it a regression, since the behaviour for an invalid handler was not defined. An alternative I considered was:

```
* Added a system check for when custom error handler cannot be imported, to
  prevent an exception from being raised. (:ticket:`30318`).
```

Or I'd be fine with any other wording you prefer.

File: django/urls/resolvers.py (hunk 1 above)
Comment (pope1ni) [reply django/urls/resolvers.py]: If you want the message of the exception, just use `str(e)`?

File: django/urls/resolvers.py [hunk 3]
```diff
@@ -405,7 +405,16 @@ def _check_custom_error_handlers(self):
...
         # which takes (request).
         for status_code, num_parameters in [(400, 2), (403, 2), (404, 2), (500, 1)]:
-            handler, param_dict = self.resolve_error_handler(status_code)
+            try:
+                handler, param_dict = self.resolve_error_handler(status_code)
+            except (ModuleNotFoundError, ViewDoesNotExist) as e:
+                path = getattr(self.urlconf_module, 'handler%s' % status_code)
+                msg = (
+                    "The custom handler{status_code} view '{path}' could not be imported."
+
```
Comment (pope1ni) [inline django/urls/resolvers.py]: Please remove this blank line.

**Comment by felixxm** [comment]
I rebased and pushed final edits. I've found one more case when exception can be raised.

File: django/urls/resolvers.py [hunk 4]
```diff
@@ -405,7 +405,15 @@ def _check_custom_error_handlers(self):
         # All handlers take (request, exception) arguments except handler500
         # which takes (request).
         for status_code, num_parameters in [(400, 2), (403, 2), (404, 2), (500, 1)]:
-            handler, param_dict = self.resolve_error_handler(status_code)
+            try:
+                handler, param_dict = self.resolve_error_handler(status_code)
+            except (ImportError, ModuleNotFoundError, ViewDoesNotExist) as e:
```
Comment (alasdairnicol) [inline django/urls/resolvers.py]: `ModuleNotFoundError` is a subclass of `ImportError`, so we could simplify this to `except (ImportError, ViewDoesNotExist)`.

**Comment by alasdairnicol** [comment]
> I rebased and pushed final edits. I've found one more case when exception can be raised.

@felixxm good catch

File: django/urls/resolvers.py (hunk 4 above)
Comment (felixxm) [reply django/urls/resolvers.py]: Right, Thanks!

**Pull request merged**